Improve reporting of unhandled errors.

twosigmajab opened the pull request:
- Log (rather than swallow) unhandled Kerberos errors.

UPDATE: removed in latest revision:

- ~Allow supplying custom 401 and 403 response bodies per call.~ 

- ~When there is an unhandled Kerberos error, send it in the body of the 403 response using the new support for custom bodies per call.~

This significantly improves debuggability when GSS API calls fail for some reason.

twosigmajab commented on wsgi_kerberos.py:
```diff
@@ -156,12 +162,13 @@ def _authenticate(self, client_token):
-            pass
+        except kerberos.GSSError as e:
+            exc = e
+            LOG.exception('Unhandled GSSError')
```
(As a reminder, `log.exception` provides the exception info in the log message automatically: https://docs.python.org/3/library/logging.html#logging.Logger.exception)

mkomitee replied:
Is the behavior the same in python2? I assume the answer is yes.

The only potential issue I see for this is that we’re going to be logging a full stack trace when this occurs, whereas in the past it was silent. Maybe we should make this opt-in?

twosigmajab replied:
> Is the behavior the same in python2? I assume the answer is yes.

[Yes](https://docs.python.org/2/library/logging.html#logging.Logger.exception). 

> The only potential issue I see for this is that we’re going to be logging a full stack trace when this occurs, whereas in the past it was silent. Maybe we should make this opt-in?

In the latest revision, I've added a `debug` param to the constructor that defaults to False. The full traceback is now only logged when this is set to True, otherwise just the exception itself is logged (with equivalent level, namely "error").

twosigmajab commented on wsgi_kerberos.py:
```diff
@@ -202,4 +209,8 @@ def custom_start_response(status, headers, exc_info=None):
             return self._unauthorized(start_response, server_token)
```
Should this use the new `extra_body` param to include some "Authorization missing" message in the 401 body?

twosigmajab replied:
Done in the latest revision.

twosigmajab commented on wsgi_kerberos.py:
```diff
@@ -202,4 +209,8 @@ def custom_start_response(status, headers, exc_info=None):
+            return self._forbidden(
+                environ,
+                start_response,
+                extra_body=b'\n\n%r' % exc
```
Note: I don't love how this composes with the existing support for custom response bodies[1], but maybe it's good enough.

[1] Rather than accepting a string or tuple in the constructor for a custom body and content-type, normalizing, and then not knowing here whether we're changing something custom that the user wants to do, the constructor could instead accept a callback for creating a custom body that we pass the extra info to, and the user can use or ignore it as they see fit. (And I'd accept that as a separate param from the custom content type, rather than having to keep packing/unpacking them into/from the 2-tuple.) But at this point I think it's better to optimize for debuggability of error-prone failure modes (e.g. forgetting to init keytabs) over some API redesign for more customizable error pages.

mkomitee replied:
I don’t think it’s a good idea to leak this information back to the client in the body by default.

twosigmajab replied:
In the latest revision, the exception is only sent back to the client when we were initialized with `debug=True`. The default is False.

twosigmajab replied:
> I don’t think it’s a good idea to leak this information back to the client in the body by default.

Based on various parts of the HTTP specs, I think this sort of thing is okay. For example, https://tools.ietf.org/html/rfc7231#section-6.5.3 says, "A server that wishes to make public why the request has been forbidden can describe that reason in the response payload (if any)." And https://tools.ietf.org/html/rfc7235#section-3.1 says, upon getting a 401 response, "the user agent SHOULD present the enclosed representation to the user, since it usually contains relevant diagnostic information."

twosigmajab commented:
Hi @mkomitee, thanks for reviewing. I believe I've addressed all your review comments. Please let me know what you think, and hope this looks better now!

propertone reviewed (changes requested):
Few comments:
- The server-side error logging and KrbError -> GSSError change looks good
- Even though it's opt-in, I don't think returning the server-side error to the client is a good idea as it could unintentionally leak server-side details. And the name "debug" certainly does not imply it. Having the server log the exception seems like it should be sufficient, no? If you control the server code enough to be able to choose the authentication library, then you should have access to the logs for debugging.

twosigmajab committed:
Log rather than swallow unhandled GSSErrors.

Also changes the code to catch GSSError rather than KrbError, which is
too broad.

twosigmajab commented:
@propertone Thanks for reviewing! Pushed a new revision with the changes you requested. Another look?

propertone reviewed (approved):
Looks good, thanks!

twosigmajab commented:
Thanks for merging!
